### PR TITLE
[SYCL-MLIR] Detect accessor range and offset

### PR DESCRIPTION
**Pull request by sommerlukas**
During raising of the host code, detect whether a constructed accessor requires the range and offset parameter.

### CODE REVIEW

**Review by victor-eds — APPROVED** [review]
Good addition :+1:

File: polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp [hunk 1]
```diff
@@ -510,10 +523,10 @@ class BufferTypeTag {
     // the type of the first parameter of the demangled constructor function
     // name.
     bool isSubBuffer = false;
-    StringRef demangled{demangledName};
-    if (auto paramStart = demangled.find('('); paramStart != StringRef::npos) {
-      if (demangled.drop_front(paramStart + 1).starts_with("sycl::_V1::buffer"))
-        isSubBuffer = true;
+    auto failureOrParams = getDemangledParameterList(constructorName);
+    if (succeeded(failureOrParams)) {
+      auto paramList = static_cast<StringRef>(*failureOrParams);
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp:528]: Should the cast be to type `DemangleResult` ? And once you have a `DemangleResult` you can call operator() to get a StringRef ?

File: polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp [hunk 2]
```diff
@@ -557,6 +570,15 @@ class AccessorTypeTag {
     llvm::SmallVector<StringRef> matches;
     bool regexMatch = accessorTemplate.match(demangledName, &matches);
 
+    bool needsRange = true;
+    bool needsID = true;
+    auto failureOrParams = getDemangledParameterList(constructorName);
+    if (succeeded(failureOrParams)) {
+      auto paramList = static_cast<StringRef>(*failureOrParams);
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp:577]: See comment at line 528

**Review by whitneywhtsang — APPROVED** [review]

File: polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp (hunk 1 above)
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp:528]: I c, ok

**Pull request merged**